Adds AngleUnit into Servo wrapper

serivesmejia opened the pull request:
# Pull Requests

Please note that we accept pull requests from anyone, but that does not mean it will be merged.

## What kind of change does this PR introduce?
* Refactor/Feature (testing pending)

## Did this PR introduce a breaking change?
* Yes, please list breaking changes

    - rotateDegrees() method renamed to rotateAngle()
    - removed constructors without min and max angle parameters, for forcing user to enter the range manually which is dependent on the servo they use (per suggestion from @dansman805)
    - changed constructor param order from max, min to min, max since it makes more sense to have that order.

__Please make sure your PR satisfies the requirements of [the contributing page](CONTRIBUTING.md)__

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -3,40 +3,79 @@
-            angle = minAngle;
 
-        setPosition((angle - minAngle) / (getAngleRange()));
+        //use local variable in case we need to convert units to degrees
+        //(remember these values are always stored as radians internally)
+        double iMaxAngle = maxAngle;
+        double iMinAngle = minAngle;
+
+        if(angleUnit == AngleUnit.DEGREES) {
+            iMaxAngle = Math.toDegrees(iMaxAngle);
+            iMinAngle = Math.toDegrees(iMinAngle);
+        }
+
+        if(angle > iMaxAngle)
+            angle = iMaxAngle;
+        else if(angle < iMinAngle)
+            angle = iMinAngle;
```
AngleUnit also has a static normalization method that clamps to `[-180, 180)` degrees or the comparable units in radians. You can easily simplify this to:
```suggestion
        //use local variable in case we need to convert units to degrees
        //(remember these values are always stored as radians internally)
        double iMaxAngle = angleUnit == AngleUnit.DEGREES ? AngleUnit.normalizeDegrees(angleUnit.toDegrees(maxAngle)) : maxAngle;
                                          
        double iMinAngle = angleUnit == AngleUnit.DEGREES ? AngleUnit.normalizeDegrees(angleUnit.toDegrees(minAngle)) : minAngle;
```

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -85,11 +129,23 @@ public double getPosition() {
     @Override
     public double getAngle() {
-        return getPosition() * getAngleRange() + minAngle;
+        //use local variable in case we need to convert units to degrees
+        //(remember this value is always stored as radians internally)
+        double iMinAngle = minAngle;
+
+        if(angleUnit == AngleUnit.DEGREES)
+            iMinAngle = Math.toDegrees(iMinAngle);
```
Just as before, you can turn this into a ternary, i.e.
```java
double iMinAngle = angleUnit == AngleUnit.DEGREES ? angleUnit.toDegrees(minAngle) : minAngle;
```

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -85,11 +129,23 @@ public double getPosition() {
     public double getAngleRange() {
-        return maxAngle - minAngle;
+        double angleRangeRadians = maxAngle - minAngle;
+        if(angleUnit == AngleUnit.DEGREES) {
+            return Math.toDegrees(angleRangeRadians);
+        } else {
+            return angleRangeRadians;
+        }
```
This can probably be a ternary in a return statement.

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -55,8 +94,17 @@ else if(position < minAngle)
+
     @Override
     public void setRange(double min, double max) {
+        if(angleUnit == AngleUnit.DEGREES) {
+            min = Math.toRadians(min);
+            max = Math.toRadians(max);
+        }
         this.minAngle = min;
         this.maxAngle = max;
```
This can be a ternary.

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -3,40 +3,79 @@
-        this.maxAngle = maxAngle;
-        this.minAngle = minAngle;
+
+        if(angleUnit == AngleUnit.DEGREES) {
+            this.maxAngle = Math.toRadians(maxAngle);
+            this.minAngle = Math.toRadians(minAngle);
+        } else {
+            this.maxAngle = maxAngle;
+            this.minAngle = minAngle;
+        }
```
This should be shortened to a ternary.

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -3,40 +3,79 @@
+        }
+
+        this.angleUnit = angleUnit;
+
```
Unneeded whitespace.

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -3,40 +3,79 @@
+    }
+
+    public SimpleServo(HardwareMap hw, String servoName, double maxAngle, double minAngle, AngleUnit angleUnit) {
+
```
Unneeded whitespace.

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -3,40 +3,79 @@
+
+        maxAngle = Math.toRadians(180);
         minAngle = 0;
+
```
Unneeded whitespace.

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -3,40 +3,79 @@
 
-    public SimpleServo(HardwareMap hw, String servoName) {
+    public SimpleServo(HardwareMap hw, String servoName, AngleUnit angleUnit) {
+
```
Unneeded whitespace.

JIceberg reviewed (changes requested):
Mostly just feedback on the formatting of the code/codestyle.

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -2,41 +2,63 @@
-        maxAngle = 180;
+        this.angleUnit = angleUnit;
+
+        maxAngle = Math.toRadians(180);
```
`Math.PI` would suffice.

JIceberg reviewed (changes requested):
I think there are still a bit of whitespace issues. Can you do Ctrl+Alt+L or whatever auto-formatting binding you have?

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/SimpleServo.java:
```diff
@@ -47,35 +69,33 @@ public void rotate(double position) {
     public void setRange(double min, double max) {
-        this.minAngle = min;
-        this.maxAngle = max;
+        this.minAngle = angleUnit == AngleUnit.DEGREES ? Math.toRadians(min) : min;
+        this.maxAngle = angleUnit == AngleUnit.DEGREES ? Math.toRadians(max) : max;
```
`Math.toRadians(double)` on degrees?

JIceberg reviewed (changes requested):
As @dansman805 was saying, the AngleUnit should be passed per-method call and not per-instance. It's up to you if you want to have two methods, one that has no AngleUnit and uses degrees by default and one that does have AngleUnit, or just have the AngleUnit.

serivesmejia commented:
Updated PR description with some other breaking changes! :)

JIceberg commented:
> some other breaking changes

ಠ_ಠ
Can you list them all for me so we can include them in a list for the version description if this gets merged for 1.2.0?

serivesmejia commented:
> ಠ_ಠ
> Can you list them all for me so we can include them in a list for the version description if this gets merged for 1.2.0?

sure, as of right now,
- breaking changes:
    - rotateDegrees() method renamed to rotateAngle()
    - removed constructors without min and max angle parameters, for forcing user to enter the range manually which is dependent on the servo they use (per suggestion from @dansman805)
    - changed constructor param order from *maxAngle, minAngle* to *minAngle, maxAngle* since it makes more sense to have that order.

- non-breaking:
   - adds an overloaded method with an AngleUnit parameter for each of the methods which take an angle parameter, in SimpleServo and ServoEx interface. methods without an AngleUnit param are defaulted as degrees, which makes this change non-breaking
   - adds a constructor taking an angle unit ONLY FOR the min and max parameters given in that constructor, like the methods, there's an overloaded constructor without the AngleUnit parameter which defaults to DEGREES
   - minAngle and maxAngle are now internally stored as radians always

JIceberg commented on core/src/main/java/com/arcrobotics/ftclib/hardware/ServoEx.java:
```diff
@@ -1,23 +1,41 @@
-     * Rotates the servo a certain number of degrees.
+     * Rotates the servo a certain angle.
+     *
+     * @param angle The desired angle of rotation
+     * @param angleUnit The unit of the angle parameter
+     */
+    void rotateAngle(double angle, AngleUnit angleUnit);
+
+    /**
+     * Rotates the servo a certain angle in degrees.
      *
-     * @param degrees The desired degrees of rotation
+     * @param angle The desired angle of rotation in degrees
      */
-    void rotateDegrees(double degrees);
+    void rotateAngle(double degrees);
```
I feel like `rotateBy` is a better name here, since it rotates the current position by a certain number of degrees or radians.

JIceberg replied:
Actually, `rotate` should be `rotateBy` and `rotateAngle` should be `rotateByAngle`. It does introduce another breaking change, though.
